File: all/704_binary_search/sol_left_close_right_close_binary_search.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left_index = 0
        right_index = len(nums) - 1

        while left_index <= right_index:
            middle_index = left_index + ((right_index - left_index) >> 1)

            if target == nums[middle_index]:
                return middle_index
            elif target > nums[middle_index]:
                left_index = middle_index + 1
            else:
                right_index = middle_index - 1

        return -1


if '__main__' == __name__:
    sol = Solution()
    testcases_list = [
        [[-1, 0, 3, 5, 9, 12], 9],
        [[-1, 0, 3, 5, 9, 12], 2]
    ]

    for testcase in testcases_list:
        print(sol.search(*testcase))

File: all/704_binary_search/sol_left_close_right_close_binary_search.c
int search(int* nums, int numsSize, int target) {
    int left_index = 0, right_index = numsSize - 1;
    int middle_index = 0;

    while (left_index <= right_index) {
        middle_index = left_index + ((right_index - left_index) >> 1);

        if (nums[middle_index] == target) {
            return middle_index;
        }
        else if (target > nums[middle_index]) {
            left_index = middle_index + 1;
        }
        else {
            right_index = middle_index - 1;
        }
    }

    return -1;
}
